tiledb/api/c_api_support/exception_wrapper/exception_wrapper.h

```diff
@@ -514,18 +514,43 @@ using ExceptionActionCtxErr = detail::ExceptionActionDetailCtxErr;
 //-------------------------------------------------------
 // Exception wrapper
 //-------------------------------------------------------
+/**
+ * Null aspect for `class CAPIFunction` has null operations for all aspects.
+ * @tparam f
+ */
+template <auto f>
+class CAPIFunctionNullAspect {
+ public:
+  template <typename... Args>
+  static void call(Args...) {
+  }
+};
+
+/**
+ * Selection struct defines the default aspect type for CAPIFunction. This class
+ * is always used with second template argument as `void`. This definition is
+ * for the general case; a specialization can override it.
+ */
+template <auto f, typename>
+struct CAPIFunctionSelector {
+  using aspect_type = CAPIFunctionNullAspect<f>;
+};
+
 /**
  * Non-specialized wrapper for implementations functions for the C API. May
  * only be used as a specialization.
  */
-template <auto f, class H>
+template <
+    auto f,
+    class H,
+    class A = typename CAPIFunctionSelector<f, void>::aspect_type>
```

Comment:
This is nice.  IIUC, the choice of `f` will bring in the right `aspect_type`.  The user can create a binding of sorts between the function that is being called and the aspect that they want to pass in -- which can be done on a case-by-case basis -- all without having to modify any of the internal code.  It's just a matter of setting up the trait (simple to do and completely outside the implementation of `CAPIFunction` -- and voila -- `CAPIFunction` can be customized to their heart's content on a function-by-function basis.  It warms my metaprogramming heart.